modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryTuplePrefix.java

```diff
@@ -73,6 +97,102 @@ public static BinaryTuplePrefix fromBinaryTuple(BinaryTuple tuple) {
         return new BinaryTuplePrefix(tuple.elementCount(), prefixBuffer);
     }
 
+    private static BinaryTuplePrefix expandTuple(int size, BinaryTuple tuple) {
+        assert size > tuple.elementCount();
+
+        var stats = new Sink() {
+            int dataBeginOffset = 0;
+            int dataEndOffset = 0;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                if (index == 0) {
+                    dataBeginOffset = begin;
+                }
+
+                dataEndOffset = end;
+            }
+        };
+
+        tuple.parse(stats);
+
+        ByteBuffer tupleBuffer = tuple.byteBuffer();
+
+        byte flags = tupleBuffer.get(0);
+        int entrySize = BinaryTupleCommon.flagsToEntrySize(flags);
+
+        ByteBuffer prefixBuffer = ByteBuffer.allocate(
+                        tupleBuffer.remaining()
+                                + (entrySize * (size - tuple.elementCount()))
+                                + Integer.BYTES)
+                .order(ORDER)
+                .put(tupleBuffer.slice().limit(stats.dataBeginOffset)); // header
+
+        int payloadEndPosition = stats.dataEndOffset - stats.dataBeginOffset;
+        for (int idx = tuple.elementCount(); idx < size; idx++) {
+            switch (entrySize) {
+                case Byte.BYTES:
+                    prefixBuffer.put((byte) payloadEndPosition);
+                    break;
+                case Short.BYTES:
+                    prefixBuffer.putShort((short) payloadEndPosition);
+                    break;
+                case Integer.BYTES:
+                    prefixBuffer.putInt(payloadEndPosition);
+                    break;
+                default:
+                    assert false;
+            }
+        }
+
+        prefixBuffer
+                .put(tupleBuffer.slice().position(stats.dataBeginOffset).limit(stats.dataEndOffset)) // payload
+                .putInt(tuple.elementCount())
+                .flip();
+
+        prefixBuffer.put(0, (byte) (flags | PREFIX_FLAG));
+
+        return new BinaryTuplePrefix(size, prefixBuffer);
+    }
+
+    private static BinaryTuplePrefix truncateTuple(int size, BinaryTuple tuple) {
+        assert size < tuple.elementCount();
+
+        var stats = new Sink() {
+            int dataBeginOffset = 0;
+            int dataEndOffset = 0;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                if (index == 0) {
+                    dataBeginOffset = begin;
+                }
+
+                if (index < size) {
+                    dataEndOffset = end;
+                }
+            }
+        };
+
+        tuple.parse(stats);
+
+        BinaryTuplePrefixBuilder builder = new BinaryTuplePrefixBuilder(size, size, stats.dataEndOffset - stats.dataBeginOffset);
+
+        tuple.parse((index, begin, end) -> {
```

Comment:
Same here:
```
for (int i = 0; i < size; i++) {
    byte[] valueBytes = tuple.bytesValue(i);
    
    builder.appendBytes(valueBytes);
}
```

Reply:
BTW, maybe we could introduce a more effective API to the BinaryTuple, because in this approach we do a lot of array copying

Reply:
answered [here](https://github.com/apache/ignite-3/pull/2728#discussion_r1370086037)

Reply:
You've answered my comment about the new API, but kept the same API, why?

Reply:
sorry, I thought it all about array copying...

if it's about iterating value by value, then it's because we can just copy prefix of header as we do in `expandTuple` method. Due to truncation, the size of element in offset map may be adjusted as well, I don't want do reimplement that logic once again.

A new API (like `BinTuple.project(int... fields):ByteBuffer`) will definitely help to make this case a lot cleaner, but the truth is, `truncateTuple` is used only in tests for now. I'm not sure if anyone will use it future, thus don't pay too much attention to this method. I believe we can revise it once it will start cause any problem

Reply:
make sense, replaced